src/lib/blockchain_snark/blockchain_snark_state.ml

```diff
@@ -86,26 +91,40 @@ let non_pc_registers_equal_var t1 t2 =
 let%snarkydef step ~(logger : Logger.t)
     ~(proof_level : Genesis_constants.Proof_level.t)
     ~(constraint_constants : Genesis_constants.Constraint_constants.t)
-    Hlist.HlistId.
-      [ previous_state_hash
-      ; (txn_snark : Transaction_snark.Statement.With_sok.Checked.t)
-      ] new_state_hash : _ Tick.Checked.t =
+    new_state_hash : _ Tick.Checked.t =
+  let new_state_hash =
+    State_hash.var_of_hash_packed (Data_as_hash.hash new_state_hash)
+  in
   let%bind transition =
     with_label __LOC__
       (exists Snark_transition.typ ~request:(As_prover.return Transition))
   in
-  let%bind previous_state, previous_state_body_hash =
+  let%bind txn_snark =
+    with_label __LOC__
+      (exists Transaction_snark.Statement.With_sok.typ
+         ~request:(As_prover.return Txn_snark) )
+  in
+  let%bind ( previous_state
+           , previous_state_hash
+           , previous_blockchain_proof_input
+           , previous_state_body_hash ) =
+    let%bind prev_state_ref =
+      with_label __LOC__
+        (exists (Typ.Internal.ref ()) ~request:(As_prover.return Prev_state))
```

Comment:
why are we using `Typ.Internal.ref ()` and not `Protocol_state.typ`?

Reply:
We need the `As_prover.Ref.t` for `previous_blockchain_proof_input` anyway, it ends up being less code to do it this way.